src/endpoint/FaabricEndpoint.cpp

```diff
@@ -1,92 +1,261 @@
+          std::bind_front(&HttpConnection::doRead, this->shared_from_this()));
+    }
+
+  private:
+    void doRead()
+    {
+        parser.body_limit(boost::none);
+        faabric::util::SystemConfig& conf = faabric::util::getSystemConfig();
+        stream.expires_after(std::chrono::seconds(conf.globalMessageTimeout));
+        beast::http::async_read(
+          stream,
+          buffer,
+          parser,
+          std::bind_front(&HttpConnection::onRead, this->shared_from_this()));
+    }
+
+    void handleRequest(faabric::util::BeastHttpRequest msg)
+    {
+        HttpRequestContext hrc{ ioc,
+                                stream.get_executor(),
+                                std::bind_front(&HttpConnection::sendResponse,
+                                                this->shared_from_this()) };
+        handler->onRequest(std::move(hrc), std::move(msg));
+    }
+
+    void onRead(beast::error_code ec, size_t bytesTransferred)
+    {
+        UNUSED(bytesTransferred);
+        if (ec == beast::http::error::end_of_stream) {
+            doClose();
+            return;
+        }
+        if (ec) {
+            SPDLOG_ERROR("Error reading an HTTP request: {}", ec.message());
+            return;
+        }
+        SPDLOG_TRACE("Read HTTP request of {} bytes", bytesTransferred);
+        handleRequest(parser.release());
+    }
+
+    void sendResponse(faabric::util::BeastHttpResponse&& response)
+    {
+        // response needs to be freed after the send completes
+        auto ownedResponse = std::make_shared<faabric::util::BeastHttpResponse>(
+          std::move(response));
+        ownedResponse->prepare_payload();
+        beast::http::async_write(stream,
+                                 *ownedResponse,
+                                 std::bind_front(&HttpConnection::onWrite,
+                                                 this->shared_from_this(),
+                                                 ownedResponse));
+    }
+
+    void onWrite(std::shared_ptr<faabric::util::BeastHttpResponse> response,
+                 beast::error_code ec,
+                 size_t bytesTransferred)
+    {
+        bool needsEof = response->need_eof();
+        response.reset();
+        UNUSED(bytesTransferred);
+        if (ec) {
+            SPDLOG_ERROR("Couldn't write HTTP response: {}", ec.message());
+            return;
+        }
+        SPDLOG_TRACE("Write HTTP response of {} bytes", bytesTransferred);
+        if (needsEof) {
+            doClose();
+            return;
+        }
+        // reset parser to a fresh object, it has no copy/move assignment
+        parser.~parser();
+        new (&parser) decltype(parser)();
+        doRead();
+    }
+
+    void doClose()
+    {
+        beast::error_code ec;
+        stream.socket().shutdown(asio::socket_base::shutdown_send, ec);
+        // ignore errors on connection closing
+    }
+};
+
+class EndpointListener : public std::enable_shared_from_this<EndpointListener>
```

Comment:
If we rename to `FaabricHttpEndpoint` we can rename this to `HttpEndpointListener`.

Can you also add a little comment on top of the class explaining what it does?